### ISSUE TITLE
Database architecture

### ISSUE THREAD

**Issue by araujo-hilquias** [taco-ide/dummy-ta#7]
We have mentioned the possible data to collect in #5 .

- Is there any other info that we need to be prepared to collect and save?
- Which would be the best architecture (tables, foreign keys, columns, etc.) to implement?
- Is there anything else to consider?

**Comment by LucasPilla** [comment]
I have some doubts regarding the requirements:
- Should be possible to have multiple `models` per `problem`?
- Are we going to organize `students` and `problems` into `classes`?

<hr>

An initial schema considering a SQL database:

- **Sessions**
	`session_id` (Primary Key, UUID)
	`username` (Foreign Key -> `Users[username]`)
	`start time` (timestamp)
	`end_time` (timestamp)

- **Users**
	`username` (Primary Key, string)
	`email`
	`password_hash`
	`role` (`teacher` or `student`)

- **Courses**
	`course_id` (Primary Key, UUID)
  	`teacher` (Foreign Key -> `Users[username]`)
	`title`
	`description`
	`created_at` (timestamp)

- **UserCourses** (relationship table)
	`username` (Foreign Key -> `User[username]`) 
  	`course_id` (Foreign Key -> `Courses[course_id]`) 

- **Problems**
	`problem_id` (Primary Key, UUID)
   	`course_id` (Foreign Key -> `Courses[course_id]`, Nullable) 
	`model_id` (Foreign Key -> `Models[model_id]`)
	`model_version` (Foreign Key -> `Models[model_version]`)
	`title`
	`description`
	`support_materials` (list of ids that identify documents in a blob storage)
	`possible_solutions`
	`created_at` (timestamp)

- **Models**
	`model_id` (Primary Key, UUID)
	`model_version` (Primary Key)
	`name`
	`description`
	`created_at` (timestamp)

- **LLM Interactions**
	`interaction_id` (Primary Key, UUID)
	`session_id` (Foreign Key -> `Sessions[session_id]`)
	`problem_id` (Foreign Key -> `Problems[problem_id]`)
	`user_prompt`
	`model_response`
	`code`
	`stdin`
	`stdout`
	`created_at` (timestamp)

**Comment by verri** [comment]
Hi Lucas

> Should be possible to have multiple models per problem?

I think not.

> Are we going to organize students and problems into classes?

I wouldn't organize students into classes, as this changes too much.

I think each problem can be associated to a course (that is a subject offered for a specific amount of time).

Students can enroll in a course, so the given problems appear in their list.

> An initial schema considering a SQL database:

About the schema, there are no test cases. Teacher writes a problem statement, possible solutions and upload support materials. So, there are no "submissions" per se. The submission is just the history of interactions (maybe we should provide a "save" button that acts like a interaction...)

**Comment by LucasPilla** [comment]
@verri I made some updates to the schema, let me know what do you think.

Changes:
- Removed **Submissions**.
- Added the following fields to **Problems**:
    - `support_materials`
    - `possible_solutions`
    - `course_id`
- Introduced two new entities:
    - **Courses**
    - **UserCourses** (to account for students being enrolled in multiple courses simultaneously).
   
Could you clarify what you mean by possible solutions?

**Comment by verri** [comment]
I mean the source code for possible solutions, e.g. "returning the n-th Fibonacci number":

```python
def fibonacci(n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    
    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, a + b
    return b
```

or 

```python
from functools import lru_cache

@lru_cache(maxsize=None)
def fibonacci(n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    return fibonacci(n - 1) + fibonacci(n - 2)
```

**Comment by LucasPilla** [comment]
Hmm, the student won't have an automatic score, like 'X tests passed,' so the idea is that the teacher will be responsible for the evaluation, right?

**Comment by verri** [comment]
Yes, exactly. It is not a "compare" system. The teacher can afterwards evaluate the "learning progress" by looking at the history (changes in the source code, questions to the LLM, etc.)

**Comment by alex81632** [comment]
> * Courses

I like the idea of courses, but I think it would be interesting to keep the relationship between course and problem optional, allowing the user to create solo problems.

**Comment by alex81632** [comment]
> * LLM Interactions

In addition to these fields, we need to send LLM the stdin and output of the code, so I think it would be interesting to include these fields.

**Comment by LucasPilla** [comment]
I updated the schema:
- We can make the `course_id` field in **Problems** nullable, so the problem doesn't have to be related to a course.
- Added `stdin` and `stdout` fields to **LLM Interactions**.
